Modules/_abc.c

```diff
@@ -0,0 +1,960 @@
+_Py_IDENTIFIER(__dict__);
+_Py_IDENTIFIER(__bases__);
+_Py_IDENTIFIER(_abc_impl);
+
+/* A global counter that is incremented each time a class is
+   registered as a virtual subclass of anything.  It forces the
+   negative cache to be cleared before its next use.
+   Note: this counter is private. Use `abc.get_cache_token()` for
+   external code. */
+static PyObject *abc_invalidation_counter;
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t iterating;
+    PyObject *data; /* The actual set of weak references. */
+    PyObject *pending; /* Pending removals collected during iteration. */
+    PyObject *in_weakreflist;
+} _guarded_set;
+
+static void
+gset_dealloc(_guarded_set *self)
+{
+    Py_DECREF(self->data);
+    Py_DECREF(self->pending);
+    if (self->in_weakreflist != NULL) {
+        PyObject_ClearWeakRefs((PyObject *) self);
+    }
+    Py_TYPE(self)->tp_free(self);
+}
+
+static PyObject *
+gset_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *data = PySet_New(NULL);
+    if (data == NULL) {
+        return NULL;
+    }
+    PyObject *pending = PyList_New(0);
+    if (pending == NULL) {
+        Py_DECREF(data);
+        return NULL;
+    }
+    _guarded_set *self = (_guarded_set *) type->tp_alloc(type, 0);
+    if (self != NULL) {
+        self->iterating = 0;
+        self->data = data;
+        self->pending = pending;
+        self->in_weakreflist = NULL;
+    }
+    return (PyObject *) self;
+}
+
+PyDoc_STRVAR(guarded_set_doc,
+"Internal weak set guarded against deletion during iteration.\n\
+Used by ABC machinery.");
+
+static PyTypeObject _guarded_set_type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    "_guarded_set",                     /*tp_name*/
+    sizeof(_guarded_set),               /*tp_size*/
+    .tp_dealloc = (destructor)gset_dealloc,
+    .tp_flags = Py_TPFLAGS_DEFAULT,
+    .tp_weaklistoffset = offsetof(_guarded_set, in_weakreflist),
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_new = gset_new,
+};
+
+/* This object stores internal state for ABCs.
+   Note that we can use normal sets for caches,
+   since they are never iterated over. */
+typedef struct {
+    PyObject_HEAD
+    _guarded_set *_abc_registry;
+    PyObject *_abc_cache; /* Normal set of weak references. */
+    PyObject *_abc_negative_cache; /* Normal set of weak references. */
+    PyObject *_abc_negative_cache_version;
+} _abc_data;
+
+static void
+abc_data_dealloc(_abc_data *self)
+{
+    Py_DECREF(self->_abc_registry);
+    Py_DECREF(self->_abc_cache);
+    Py_DECREF(self->_abc_negative_cache);
+    Py_DECREF(self->_abc_negative_cache_version);
+    Py_TYPE(self)->tp_free(self);
+}
+
+static PyObject *
+abc_data_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *registry = NULL, *cache = NULL, *negative_cache = NULL;
+    registry = gset_new(&_guarded_set_type, NULL, NULL);
+    cache = PySet_New(NULL);
+    negative_cache = PySet_New(NULL);
+    if (!registry || !cache || !negative_cache) {
```

Comment:
Check `gset_new` for NULL return, then call `PySet_New` and check, then call another `PySet_New` and check